src/runtime/objmodel.cpp

```diff
@@ -3366,8 +3387,17 @@ void rearrangeArguments(ParamReceiveSpec paramspec, const ParamNames* param_name
     for (int i = 0; i < paramspec.num_args - paramspec.num_defaults; i++) {
         if (params_filled[i])
             continue;
-        raiseExcHelper(TypeError, "%s() takes exactly %d arguments (%ld given)", func_name, paramspec.num_args,
-                       argspec.num_args + argspec.num_keywords + varargs.size());
+        long given = argspec.num_args + argspec.num_keywords + varargs.size();
+        int takes = paramspec.num_args;
+        if (!(paramspec.takes_varargs || paramspec.takes_kwargs)) {
+            raiseExcHelper(TypeError, "%s() takes exactly %d argument%s (%ld given)", func_name, takes,
+                           takes != 1 ? "s" : "", given);
+        } else {
+            if (given < takes) {
+                raiseExcHelper(TypeError, "%s() takes at least %d argument%s (%ld given)", func_name, takes,
+                               takes != 1 ? "s" : "", given);
+            }
+        }
```

Comment:
I don't think that's quite what's going on here -- this block is checking to make sure that all of the positional arguments received values.  Whether we passed more values into `*args` or `**kw` shouldn't matter.  Here's an example:
```python
def f(a, **kw):
  print "called function"
f(b=1, c=2)
```
With this change we will now allow the function call to succeed, when instead we should be throwing an exception.